Add additional support for parachute

amy-chen-skydio opened the pull request:
- Add messages and flags to report parachute status and information
- Add command `MAV_CMD_SET_PARACHUTE_ARM` to arm parachute with additional triggering methods

amy-chen-skydio committed:
Add additional support for parachutes
 - Add messages to report parachute status and information
 - Add commands to arm with additional trigger methods and to shutdown the module

amy-chen-skydio committed:
Update common.xml
 - Add `MAV_CMD_SET_PARACHUTE_ARM_ALTI` for setting the ATS arm altitude
 - Add `PARACHUTE_ERROR_FLAGS_GS_ERROR` and `PARACHUTE_ERROR_FLAGS_SUBSYSTEM_ERROR` in `PARACHUTE_ERROR_FLAGS`

amy-chen-skydio committed:
Update common.xml

julianoes commented on message_definitions/v1.0/common.xml:
```diff
@@ -7532,6 +7637,26 @@
+      <field type="uint8_t" name="deployment_status" enum="PARACHUTE_DEPLOYMENT_FLAGS" display="bitmask">Parachute deployment_status</field>
+      <field type="uint8_t" name="ats_arm_altitude" units="m">Parachute ATS auto-arming/disarming altitude in meters</field>
+    </message>
+    <message id="431" name="PARACHUTE_INFORMATION">
+      <description>Parachute information</description>
+      <field type="char[16]" name="serial_number">Serial number in ASCII characters, null terminated. All 0: field not provided.</field>
+      <field type="uint32_t" name="main_system_fw_version">Parachute main system firmware version, encoded as: (Dev &amp; 0xff) &lt;&lt; 24 | (Patch &amp; 0xff) &lt;&lt; 16 | (Minor &amp; 0xff) &lt;&lt; 8 | (Major &amp; 0xff). Use 0 if not known.</field>
+      <field type="uint32_t" name="main_system_hw_version">Parachute main system hardware version, encoded as: (Dev &amp; 0xff) &lt;&lt; 24 | (Patch &amp; 0xff) &lt;&lt; 16 | (Minor &amp; 0xff) &lt;&lt; 8 | (Major &amp; 0xff). Use 0 if not known.</field>
+      <field type="uint32_t" name="subsystem_fw_version">Parachute subsystem firmware version, encoded as: (Dev &amp; 0xff) &lt;&lt; 24 | (Patch &amp; 0xff) &lt;&lt; 16 | (Minor &amp; 0xff) &lt;&lt; 8 | (Major &amp; 0xff). Use 0 if not known.</field>
+      <field type="uint32_t" name="subsystem_hw_version">Parachute subsystem hardware version, encoded as: (Dev &amp; 0xff) &lt;&lt; 24 | (Patch &amp; 0xff) &lt;&lt; 16 | (Minor &amp; 0xff) &lt;&lt; 8 | (Major &amp; 0xff). Use 0 if not known.</field>
+      <field type="char[11]" name="parachute_packed_date">Parachute packed date (DD/MM/YYYY) in ASCII characters, 0 terminated. All 0: field not provided.</field>
+      <field type="char[11]" name="manufacture_date">Manufacture date (DD/MM/YYYY) in ASCII characters, 0 terminated. All 0: field not provided.</field>
+    </message>
```
Instead of having these INFORMATION messages for every type of component, we try to generalize it using generic components. The message that could be is currently in development.xml:
https://mavlink.io/en/messages/development.html#COMPONENT_INFORMATION_BASIC

As soon as there are implementations published, we can move it to common.xml.

amy-chen-skydio replied:
I have removed `PARACHUTE_INFORMATION`. I added `manufacture_date` to `COMPONENT_INFORMATION_BASIC` since this is generic and can apply to any components?

julianoes commented on message_definitions/v1.0/common.xml:
```diff
@@ -7532,6 +7637,26 @@
+      <description>Parachute system status</description>
+      <field type="uint32_t" name="time_boot_ms" units="ms">Uptime in ms</field>
+      <field type="uint32_t" name="error_status" enum="PARACHUTE_ERROR_FLAGS" display="bitmask">Errors</field>
+      <field type="uint8_t" name="battery_status" units="%">Parachute battery percentage</field>
```
The battery status could probably just be [BATTERY_STATUS](https://mavlink.io/en/messages/common.html#BATTERY_STATUS) from the parachute component.

hamishwillee replied:
HOwdy @amy-chen-skydio - some questions here for you.

amy-chen-skydio replied:
Sorry for the delay, I have removed `battery_status` from `PARACHUTE_STATUS` in the latest change

julianoes reviewed (changes requested):
Thanks for the pull request!

Now, we're trying to avoid having specific messages for every component type and try to cover it with more generic component messages. For specific configuration, we can then use parameters. I have made a few inline comments around that.

The risk with specific configuration flags is that every new device/implementation of a parachute system might have slightly different settings and require more flags and status fields.

amy-chen-skydio committed:
Update common.xml

- Remove `MAV_CMD_DO_PARACHUTE_SHUTDOWN` and use `MAV_CMD_PREFLIGHT_REBOOT_SHUTDOWN` instead
- Remove `battery_status` from `PARACHUTE_STATUS` and use `BATTERY_STATUS` instead
- Remove `PARACHUTE_INFORMATION` and use `COMPONENT_INFORMATION_BASIC` instead
    - Add `manufacture_date` to `COMPONENT_INFORMATION_BASIC`
    - Move `parachute_packed_date` to `PARACHUTE_STATUS`

amy-chen-skydio commented on message_definitions/v1.0/common.xml:
```diff
@@ -1764,6 +1764,18 @@
       </entry>
+      <entry value="215" name="MAV_CMD_DO_PARACHUTE_SHUTDOWN" hasLocation="false" isDestination="false">
+        <description>Command to shutdown the parachute module</description>
+      </entry>
```
I have removed this and will just use `MAV_CMD_PREFLIGHT_REBOOT_SHUTDOWN` instead

hamishwillee commented:
@julianoes I understand @amy-chen-skydio has modified/responded to your comments. Can you please recheck?

julianoes commented on message_definitions/v1.0/development.xml:
```diff
@@ -466,6 +466,7 @@
       <field type="char[24]" name="software_version">Software version. The recommended format is SEMVER: 'major.minor.patch'  (any format may be used). The field must be zero terminated if it has a value. The field is optional and can be empty/all zeros.</field>
       <field type="char[24]" name="hardware_version">Hardware version. The recommended format is SEMVER: 'major.minor.patch'  (any format may be used). The field must be zero terminated if it has a value. The field is optional and can be empty/all zeros.</field>
       <field type="char[32]" name="serial_number">Hardware serial number. The field must be zero terminated if it has a value. The field is optional and can be empty/all zeros.</field>
+      <field type="char[11]" name="manufacture_date">Manufacture date (DD/MM/YYYY) in ASCII characters, 0 terminated. All 0: field not provided.</field>
```
I would probably say UTC time in seconds instead of a verbose date. Also, I wonder if this should be an extension or not.

hamishwillee replied:
@julianoes Or if we are going to date it them IMO this should be a `char[9]` - possibly as YYYYMMDD.

potaito replied:
Wouldn't a UINT64 unix timestamp give us more resolution with fewer bytes (8)?

Plus unix timestamp is a recognized international format. DD/MM/YYYY on the other hand is a localized format, not used in all countries.

julianoes replied:
@potaito exactly!

hamishwillee replied:
The only reason for a stamp in this format is human readability, which need not be a deciding factor. This sounds generally useful, and as there are concerns with this PR, let's add it separately https://github.com/mavlink/mavlink/pull/2078

amy-chen-skydio replied:
Agreed and will remove this change in this PR

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -1764,6 +1764,15 @@
+        <param index="2" label="Bitmask" enum="PARACHUTE_ARM_FLAGS">Bitmask of arm flags (in param1) that should be modified</param>
+      </entry>
+      <entry value="216" name="MAV_CMD_SET_PARACHUTE_ARM_ALTI" hasLocation="false" isDestination="false">
+        <description>Command to set the parachute's automatic trigger system arming altitude in meters. </description>
```
We'd need to clean up the spurious whitespace - such as that after the full stop.

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -1764,6 +1764,15 @@
       </entry>
+      <entry value="215" name="MAV_CMD_SET_PARACHUTE_ARM" hasLocation="false" isDestination="false">
+        <description>Command to arm/disarm parachute module</description>
+        <param index="1" label="Arm Flags" enum="PARACHUTE_ARM_FLAGS">Arm Flags</param>
+        <param index="2" label="Bitmask" enum="PARACHUTE_ARM_FLAGS">Bitmask of arm flags (in param1) that should be modified</param>
```
So how does this work - I mean the enum is the same in both so why can't you just have one of these and set the value to 0 or 1 for the things you want turned on and off? 

Or is this so you don't need to know the other arm flags - you just set what you want to arm (.e.g feature A) and the bitmask, and all the other items stay the same?

amy-chen-skydio replied:
Yes, that's exactly it. With both, we will be able to just arm/disarm whatever features you want without while the other items just stay the same.

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -1764,6 +1764,15 @@
         <param index="6">Empty</param>
         <param index="7">Empty</param>
       </entry>
+      <entry value="215" name="MAV_CMD_SET_PARACHUTE_ARM" hasLocation="false" isDestination="false">
```
Shouldn't the arming / disarming be done through the message
https://mavlink.io/en/messages/common.html#MAV_CMD_COMPONENT_ARM_DISARM

and the flags turned into parameters, as @julianoes suggested?

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -1764,6 +1764,15 @@
+        <param index="1" label="Arm Flags" enum="PARACHUTE_ARM_FLAGS">Arm Flags</param>
+        <param index="2" label="Bitmask" enum="PARACHUTE_ARM_FLAGS">Bitmask of arm flags (in param1) that should be modified</param>
+      </entry>
+      <entry value="216" name="MAV_CMD_SET_PARACHUTE_ARM_ALTI" hasLocation="false" isDestination="false">
```
As far as I can tell, MAVLink tries to only have SET_ messages for things that are expected to change on every flight, maybe even multiple times, which is why parameters are unsuited. Example:
- [MAV_CMD_SET_CAMERA_MODE](https://mavlink.io/en/messages/common.html#MAV_CMD_SET_CAMERA_MODE)
- [MAV_CMD_SET_CAMERA_ZOOM](https://mavlink.io/en/messages/common.html#MAV_CMD_SET_CAMERA_ZOOM)
- [MAV_CMD_SET_CAMERA_FOCUS](https://mavlink.io/en/messages/common.html#MAV_CMD_SET_CAMERA_FOCUS)

Is the parachute arming altitude something that the user is going to modify before every flight? And what's the reference frame for MAV_CMD_SET_PARACHUTE_ARM_ALTI, above sea level or above takeoff / home? I assume the latter, which means that the setting should not change often, meaning that parameters would also work for this.

amy-chen-skydio replied:
`MAV_CMD_SET_PARACHUTE_ARM_ALTI` has been removed in the latest commit

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -7532,6 +7634,16 @@
+      <field type="uint32_t" name="time_boot_ms" units="ms">Uptime in ms</field>
+      <field type="uint32_t" name="error_status" enum="PARACHUTE_ERROR_FLAGS" display="bitmask">Errors</field>
+      <field type="uint8_t" name="arm_status" enum="PARACHUTE_ARM_FLAGS" display="bitmask">Parachute arming status</field>
+      <field type="uint8_t" name="charge_status">Boolean indicating charging (1) or not (0)</field>
```
`charge_status` does not seem specific to parachutes. Shouldn't the parachute publish [BATTERY_STATUS](https://mavlink.io/en/messages/common.html#BATTERY_STATUS) as mentioned in the previous comments instead?

hamishwillee replied:
or BATTERY_STATUS_V2 would be better, though not yet supported by anyone (see development.xml). It uses most of the same fields.

jenke-skydio replied:
I would advocate for `BATTERY_STATUS` here since we only battery percentage which can be covered by `battery_remaining` and a Boolean indicating charging (1) or not (0) which can be covered by `charge_state`

hamishwillee replied:
FWIW that makes sense, since it is understood.

BATTERY_STATUS_V2 provides the current field (positive or negative) to indicate charging. It's a much better message. FYI we are also proposing FUEL_STATUS which can be uses as a synonym which uses the same approach https://github.com/mavlink/mavlink/pull/2107

amy-chen-skydio replied:
Removed `charge_status` in the latest commit and will use `BATTERY_STATUS` instead

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -7532,6 +7634,16 @@
+      <field type="uint8_t" name="arm_status" enum="PARACHUTE_ARM_FLAGS" display="bitmask">Parachute arming status</field>
+      <field type="uint8_t" name="charge_status">Boolean indicating charging (1) or not (0)</field>
+      <field type="uint8_t" name="deployment_status" enum="PARACHUTE_DEPLOYMENT_FLAGS" display="bitmask">Parachute deployment_status</field>
+      <field type="uint8_t" name="ats_arm_altitude" units="m">Parachute ATS auto-arming/disarming altitude in meters</field>
```
I don't think Automatic Trigger System (ATS) is a common abbreviation.

```suggestion
      <field type="uint8_t" name="ats_arm_altitude" units="m">Parachute Automatic Trigger System (ATS) auto-arming/disarming altitude in meters</field>
```

julianoes replied:
Good point!

amy-chen-skydio replied:
Made the change in the latest commit!

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -4986,6 +4995,99 @@
+        <description>There is an error with the parachute external power source</description>
+      </entry>
+      <entry value="2048" name="PARACHUTE_ERROR_FLAGS_OBC_CONNECTION_ERROR">
+        <description>There is an error with the parachute and OBC (onboard computer) connection</description>
```
- "onboard" is typically used for computers inside of the component. For instance a camera that has its own compute unit that can do more than just store a photo on an SD card.
- "offboard" was historically used for a standalone computer that is not the autopilot.

That's why I'm not sure what the intended meaning in this PR is.

potaito replied:
@jenke-skydio how is the connection to the offboard computer defined? How should a parachute monitor its state? Is the parachute expected to look for heartbeat messages that have the [MAV_TYPE_ONBOARD_CONTROLLER](https://mavlink.io/en/messages/common.html#MAV_TYPE_ONBOARD_CONTROLLER) type?

jenke-skydio replied:
connection to offboard computer is optional, when its there it must send `HEARTBEAT` and optionally `SYSTEM_TIME`

hamishwillee replied:
There isn't really a place in in xml file for this kind of information. I expect you'll need to create a parachute service doc that explains the models and use cases that are addressed.

jenke-skydio replied:
yes 100%.  We plan to write a up parachute microservice documentation [in the MAVLink devguide](https://github.com/mavlink/mavlink-devguide/tree/master/en/services)

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -4986,6 +4995,99 @@
+        <description>Parachute deployed by geofence method</description>
+      </entry>
+    </enum>
+    <enum name="PARACHUTE_ERROR_FLAGS" bitmask="true">
```
Most of these errors are already captured in [SYS_STATUS](https://mavlink.io/en/messages/common.html#SYS_STATUS), which is also used by the autopilot to communicate issues with Baro, IMU, GPS etc.

potaito replied:
Fair enough!

amy-chen-skydio committed:
Update common.xml

Address PR feedback and test failures
 - Move parachute `MAV_CMD` enums since already in-use by ardupilotmega
 - Clean up spurious whitespace
 - ATS (automatic trigger system) isn't a common abbreviation
 - Removed manufacture date addition from this PR

jenke-skydio commented on message_definitions/v1.0/common.xml:
```diff
@@ -4986,6 +4995,99 @@
+      <entry value="4" name="PARACHUTE_ARM_FLAGS_MANUAL_TRIGGER_DEVICE">
+        <description>Arming parachute's manual trigger device method</description>
+      </entry>
+      <entry value="8" name="PARACHUTE_ARM_FLAGS_OBC_TRIGGER_METHOD">
```
this is intended to be the drone's onboard computer

hamishwillee replied:
@potaito Yes, they are all used interchangeably. 

Historically Offboard and Companion computer, and more recently mission computer, were used to differentiate the general computer from the FC ("the board") on a drone. Onboard computer is used for the same thing, but in this case "board" refers to the fact that the computer travels on the drone.

In future we'll see more "boards" where there is no visible separation between FC and companion, so we do need to think of a consistent terminology to refer to the bit that runs the RTOS/Flight stack, and the bit that provides general computing. 

For now, I'd say Offboard is the most commonly used.  For this case, in my suggestion I've renamed the flag to be "MAVLink".

amy-chen-skydio replied:
Renamed and updated the description per [suggestion](https://github.com/mavlink/mavlink/pull/2063#discussion_r1577246968)

hamishwillee commented:
@potaito @julianoes There has been a round of changes to this. Can you please re-review and confirm that your concerns have been addressed, and highlight any that have not.

@amy-chen-skydio Thanks for your patience. When this converges, I have the same comments to this as in this comment for the illuminators PR (how to validate/documentation required): https://github.com/mavlink/mavlink/pull/2047#issuecomment-1944819358

jenke-skydio reviewed (commented):
Addressed all open comments.  Please take a look and Amy can make changes accordingly

hamishwillee commented:
@julianoes @potaito Can you please respond to @jenke-skydio 's comments?

amy-chen-skydio commented:
> @julianoes @potaito Can you please respond to @jenke-skydio 's comments?

@potaito Wanted to follow up to see if you've had a chance to look over @jenke-skydio's comments. Thank you!

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -4986,6 +4995,99 @@
+        <description>Reserved for future use.</description>
+      </entry>
+      <entry value="64" name="PARACHUTE_ARM_FLAGS_AUTO_ARM_ATS">
+        <description>Arming parachute's auto-arming its automatic trigger system capability</description>
```
No, I still don't get it :sweat_smile: 

> auto-arm mode ... is armed

Is the auto-arm mode enabled? Is that what's intended?

jenke-skydio replied:
Yes, this is enabling the auto-arm mode.  How about: "Enabling parachute automatic trigger system's (ATS) auto-arm mode".

When parachute ATS is in auto-arm mode, for example, a vehicle may takeoff with the ATS disarmed then later detect that the vehicle has gone over a certain altitude threshold and automatically arm.  This would allow the ATS to automatically deploy the chute if frefall was detected.  I think we can make this pretty clear with a state diagram in the MAVLink devguide when we add some documentation for this microservice

potaito replied:
Yes, this is much better:
> Enabling parachute automatic trigger system's (ATS) auto-arm mode

Than this:
> Arming parachute's auto-arming its automatic trigger system capability

:+1:

hamishwillee replied:
I have simplified (I hope) the whole thing in https://github.com/mavlink/mavlink/pull/2063/files#r1577246968

> When parachute ATS is in auto-arm mode, for example, a vehicle may takeoff with the ATS disarmed then later detect that the vehicle has gone over a certain altitude threshold and automatically arm. 

If the ATS is doing an altitude check then it could simply decide it won't trigger/fully arm until until it reaches the desired altitude - so you don't need an extra state. After all, you'd have to configure the height to trigger and so on anyway. 
If it is the Vehicle doing the altitude check, then  the vehicle can get to the specific altitude and enable ATS. 

Either way, I don't see why you need it. Though I guess it is more flexible than the idea of `PARACHUTE_ARM_FLAGS_AUTO_ARM_ATS_AT_ALTITUDE

jenke-skydio replied:
The aim here is to allow the parachute system to automatically detect the altitude threshold and arm rather than the flight controller doing it.  I generally agree that in most cases you may want to have the flight controller deal with this logic.

hamishwillee replied:
So is the disarm case because you might want to turn off ATS on the takeoff or landing, but not the converse?

jenke-skydio replied:
yes

amy-chen-skydio replied:
Renamed and updated the description per [suggestion](https://github.com/mavlink/mavlink/pull/2063#discussion_r1577246968)

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -4986,6 +4995,99 @@
+        <description>There is an error with the parachute IMU</description>
+      </entry>
+      <entry value="4" name="PARACHUTE_ERROR_FLAGS_RF_ERROR">
+        <description>There is an error with the parachute RF</description>
```
What's "RF" ?

jenke-skydio replied:
this is referring to the radio, we can change to `PARACHUTE_ERROR_FLAGS_RADIO_ERROR` and `There is an error with the parachute radio`

hamishwillee replied:
There is still a question here about what does a parachute need a radio for. I assume this is a parachute connection for MAVLink? What if the channel is USB? If it is a MAVLink channel, how will this be sent when the link is dead. Note, I have not read this thoroughly, so perhaps answered elsewhere.

jenke-skydio replied:
its common for a parachute system to consist of a vehicle integrated module with an optional radio as well as an optional transmitter to be used on ground by a visual observer to deploy the chute (this is intended to align with emerging standards such as [ASTM-F3322](https://www.astm.org/f3322-22.html)).  While it is feasible that MAVLink is used for this link, it is not required.  In the case where the system is more of a self-contained system, this error would indicate a radio error internal to the system such that it would be feasible for a the vehicle or ground station to display some status or take some action

potaito replied:
That makes sense. Parachute, integrated system with its own radio :+1:

hamishwillee replied:
So is this what you mean by the manual trigger mode?

jenke-skydio replied:
Yes.  A manual trigger could come from the optional transmitter described above or could also come from the vehicle ground station via UI button and passed along from ground station to vehicle to parachute system via MAVLink

hamishwillee replied:
```suggestion
        <description>There is an error with the parachute RF (used for manual control)</description>
```

amy-chen-skydio replied:
Updated the description per suggestion above

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -4986,6 +4995,99 @@
+        <description>Arming parachute's auto-disarming its automatic trigger system capability</description>
+      </entry>
+    </enum>
+    <enum name="PARACHUTE_DEPLOYMENT_FLAGS" bitmask="true">
```
Just double-checking: `PARACHUTE_DEPLOYMENT_FLAGS` is 0 until the parachute pops, at which point it will have the flag describing the reason for the popping, right?

If that's the case, it should be an enum.

jenke-skydio replied:
i believe that is correct (CC @amy-chen-skydio )

amy-chen-skydio replied:
Updated to be used as enum instead of flags

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -4986,6 +4995,99 @@
+    <enum name="PARACHUTE_ARM_FLAGS" bitmask="true">
+      <description>Parachute module arming flags (bitmap, 0 means no arming enabled)</description>
+      <entry value="1" name="PARACHUTE_ARM_FLAGS_DRONE_TRIGGER_METHOD">
+        <description>Arming parachute's drone trigger method</description>
```
This means a control signal - PWM or whatever?

jenke-skydio replied:
This is intended to be a MAVLink command originated from the flight controller.   It is meant to be differentiated from the "offboard computer".  Seems OK to allow other types of signals (PWM, etc) but would be good to additionally call out MAVLink

hamishwillee replied:
Hmmm. So the drone looks at the MAV_TYPEs of incoming triggers and uses them to determine if the command is from the FC or the Companion computer, or a GCS (say)? Sounds complicated. If you mean this though, we should say it, and replace my name for onboard/MAVLink.

hamishwillee replied:
So how does it know it's dealing with a geofence trigger? Is this a MAVLink geofence, or an internal one, or something else?

jenke-skydio replied:
> Hmmm. So the drone looks at the MAV_TYPEs of incoming triggers and uses them to determine if the command 

is from the FC or the Companion computer, or a GCS (say)?
Yes.  I can get more detail on this but i believe the idea is around both control of configuration such that various triggers could be enabled & disabled in different scenarios as well as being able to log & audit the source of a trigger event


> So how does it know it's dealing with a geofence trigger? Is this a MAVLink geofence, or an internal one, or something else?

I need to check on this -> @amy-chen-skydio do you know?

jenke-skydio replied:
geofence is intended to leverage MAVLink mission protocol to transmit geofence data to the parachute.  The parachute would then automatically trigger based on its internal GPS system.  Again, the goal here would be for a parachute system to be able to operate autonomously even if a vehicle were not operating correctly

hamishwillee replied:
I've updated https://github.com/mavlink/mavlink/pull/2063/files#r1577246968 to say this.

But probably need to say more - in particular in accompanying docs. In particular, I assume:
1. Parachute gets geofence from vehicle, not GCS.
2. Parachute [monitors for changes in the plan](https://mavlink.io/en/services/mission.html#detecting-missionplan-changes) and minimally gets the latest on boot (if opaque id not supported).

jenke-skydio replied:
yes, both are correct: geofence comes from vehicle and should be updated on boot and both to be included on microservices documentation (fyi @amy-chen-skydio )

hamishwillee replied:
@amy-chen-skydio FYI, I'm eyes off this until there are updates.

amy-chen-skydio replied:
@hamishwillee updated per [suggestion](https://github.com/mavlink/mavlink/pull/2063#discussion_r1577246968)

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -4986,6 +4995,99 @@
         <description>Mission has executed all mission items.</description>
       </entry>
     </enum>
+    <enum name="PARACHUTE_ARM_FLAGS" bitmask="true">
+      <description>Parachute module arming flags (bitmap, 0 means no arming enabled)</description>
+      <entry value="1" name="PARACHUTE_ARM_FLAGS_DRONE_TRIGGER_METHOD">
+        <description>Arming parachute's drone trigger method</description>
+      </entry>
+      <entry value="2" name="PARACHUTE_ARM_FLAGS_AUTOMATIC_TRIGGER_SYSTEM">
+        <description>Arming parachute's automatic trigger system</description>
+      </entry>
+      <entry value="4" name="PARACHUTE_ARM_FLAGS_MANUAL_TRIGGER_DEVICE">
+        <description>Arming parachute's manual trigger device method</description>
+      </entry>
+      <entry value="8" name="PARACHUTE_ARM_FLAGS_OBC_TRIGGER_METHOD">
+        <description>Arming parachute's onboard computer trigger method</description>
+      </entry>
+      <entry value="16" name="PARACHUTE_ARM_FLAGS_GEOFENCE_TRIGGER_METHOD">
+        <description>Arming parachute's geofence trigger method</description>
+      </entry>
+      <entry value="32" name="PARACHUTE_ARM_FLAGS_RESERVED">
+        <description>Reserved for future use.</description>
+      </entry>
+      <entry value="64" name="PARACHUTE_ARM_FLAGS_AUTO_ARM_ATS">
+        <description>Arming parachute's auto-arming its automatic trigger system capability</description>
+      </entry>
+      <entry value="128" name="PARACHUTE_ARM_FLAGS_AUTO_DISARM_ATS">
+        <description>Arming parachute's auto-disarming its automatic trigger system capability</description>
+      </entry>
+    </enum>
```
A suggestion - note, query on what FC case means. Also why a manual trigger? Why do you need that?


```suggestion
    <enum name="PARACHUTE_TRIGGER_FLAGS" bitmask="true">
      <description>Enable/disable parachute trigger sources and modes</description>
      <entry value="1" name="PARACHUTE_TRIGGER_FLAGS_FC">
        <description>Flight controller trigger (e.g. PWM, DroneCan, RC Control)</description>
      </entry>
      <entry value="2" name="PARACHUTE_TRIGGER_FLAGS_ATS">
        <description>Automatic trigger system (ATS)</description>
      </entry>
      <entry value="4" name="PARACHUTE_TRIGGER_FLAGS_MANUAL">
        <description>Manual trigger</description>
      </entry>
      <entry value="8" name="PARACHUTE_TRIGGER_FLAGS_MAVLINK">
        <description>MAVLink trigger (offboard computer)</description>
      </entry>
      <entry value="16" name="PARACHUTE_TRIGGER_FLAGS_GEOFENCE">
        <description>Geofence trigger</description>
      </entry>
      <entry value="32" name="PARACHUTE_TRIGGER_FLAGS_RESERVED">
        <description>Reserved for future use.</description>
      </entry>
      <entry value="64" name="PARACHUTE_TRIGGER_FLAGS_ATS_AUTO_ARM">
        <description>Auto-arming of parachute automatic trigger system (ATS)</description>
      </entry>
      <entry value="128" name="PARACHUTE_TRIGGER_FLAGS_ATS_AUTO_DISARM">
        <description>Auto-disarming of parachute automatic trigger system (ATS)</description>
      </entry>
    </enum>
```

jenke-skydio replied:
this seems reasonable (FYI @amy-chen-skydio ).  I would add an additional note to to include `MAVLink` in addition to `PWM, DroneCan, RC Control` in the description of `PARACHUTE_TRIGGER_FLAGS_FC`

Also see discussion above about `PARACHUTE_ERROR_FLAGS_RADIO_ERROR` where we talked about the manual trigger

amy-chen-skydio replied:
Updated per above suggestion in latest commit

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -4986,6 +4995,99 @@
+        <description>Arming parachute's auto-disarming its automatic trigger system capability</description>
+      </entry>
+    </enum>
+    <enum name="PARACHUTE_DEPLOYMENT_FLAGS" bitmask="true">
+      <description>Parachute module deployment flags (bitmap, 0 means no deployment)</description>
+      <entry value="1" name="PARACHUTE_DEPLOYMENT_FLAGS_DRONE_TRIGGER_METHOD">
+        <description>Parachute deployed by drone</description>
+      </entry>
+      <entry value="2" name="PARACHUTE_DEPLOYMENT_FLAGS_AUTOMATIC_TRIGGER_SYSTEM">
+        <description>Parachute deployed by automatic trigger system</description>
+      </entry>
+      <entry value="4" name="PARACHUTE_DEPLOYMENT_FLAGS_MANUAL_TRIGGER_DEVICE">
+        <description>Parachute deployed by manual trigger device</description>
+      </entry>
+      <entry value="8" name="PARACHUTE_DEPLOYMENT_FLAGS_OBC_TRIGGER_METHOD">
+        <description>Parachute deployed by onboard computer</description>
+      </entry>
+      <entry value="16" name="PARACHUTE_DEPLOYMENT_FLAGS_GEOFENCE">
+        <description>Parachute deployed by geofence method</description>
+      </entry>
+    </enum>
```
I agree with @potaito - should be an enum if only one can be true.

```suggestion
    <enum name="PARACHUTE_DEPLOYMENT_TRIGGER" bitmask="true">
      <description>Parachute deployment trigger source</description>
      <entry value="0" name="PARACHUTE_DEPLOYMENT_TRIGGER_NONE">
        <description>None (not triggered)</description>
      </entry>
      <entry value="1" name="PARACHUTE_DEPLOYMENT_TRIGGER_DRONE">
        <description>Drone</description>
      </entry>
      <entry value="2" name="PARACHUTE_DEPLOYMENT_TRIGGER_ATS">
        <description>Automatic trigger system (ATS)</description>
      </entry>
      <entry value="3" name="PARACHUTE_DEPLOYMENT_TRIGGER_MANUAL">
        <description>Manual</description>
      </entry>
      <entry value="4" name="PARACHUTE_DEPLOYMENT_TRIGGER_MAVLINK">
        <description>MAVLink (offboard computer)</description>
      </entry>
      <entry value="5" name="PARACHUTE_DEPLOYMENT_TRIGGER_GEOFENCE">
        <description>Geofence</description>
      </entry>
    </enum>
```

jenke-skydio replied:
this seems reasonable (CC @amy-chen-skydio )

amy-chen-skydio replied:
Updated per above suggestion

amy-chen-skydio committed:
Update common.xml

Address PR feedback:
 - Remove `MAV_CMD_SET_PARACHUTE_ARM_ALTI`, using params instead
 - Rename `PARACHUTE_ARM_FLAGS` to `PARACHUTE_TRIGGER_FLAGS` and update descriptions
 - Rename `PARACHUTE_DEPLOYMENT_FLAGS` to `PARACHUTE_DEPLOYMENT_TRIGGER`, update descriptions, and use it as enum instead of bit flags
 - Update descriptions in `PARACHUTE_ERROR_FLAGS`
 - Remove `charge_status` from `PARACHUTE_STATUS`, using `BATTERY_STATUS` instead

New items added for safety and regulatory purposes:
 - `PARACHUTE_SAFETY_FLAGS` and FTS pre-checking

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -1873,6 +1873,11 @@
         <param index="7">Empty</param>
       </entry>
+      <entry value="225" name="MAV_CMD_SET_PARACHUTE_ARM" hasLocation="false" isDestination="false">
+        <description>Command to arm/disarm parachute module</description>
```
```suggestion
        <description>Command to arm/disarm parachute module trigger sources.</description>
```

hamishwillee commented:
@julianoes @potaito I believe that @amy-chen-skydio has addressed all your concerns. To my mind the descriptions for each trigger source are now clear, and this makes sense.

Are you OK with it?

hamishwillee commented:
@amy-chen-skydio Assuming that everyone is happy with this, we'll need similar tests/validations as for the illuminators to get this in.

amy-chen-skydio committed:
Update message_definitions/v1.0/common.xml

Update `MAV_CMD_SET_PARACHUTE_ARM` description

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>There is an error with the parachute and FC (flight controller) connection</description>
+      </entry>
+      <entry value="32" name="PARACHUTE_ERROR_FLAGS_EFTS_CONNECTION_ERROR">
+        <description>There is an error with the parachute electrical/external FTS (flight termination system)</description>
```
Why "electrical/external" FTS? It's always electrical, isn't it? And can this refer to internal and external Flight Termination Systems at the same time?

amy-chen-skydio replied:
I've updated the description. This should refer to Electrical Flight Termination System

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>There is an error with the parachute pod</description>
+      </entry>
+      <entry value="128" name="PARACHUTE_ERROR_FLAGS_EFTS_DIAGNOSE">
+        <description>Parachute electrical/external FTS (flight termination system) diagnosis failed</description>
```
```suggestion
        <description>Parachute electrical/external Flight Termination System (FTS) diagnosis failed</description>
```

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>There is an error with the parachute external power source</description>
+      </entry>
+      <entry value="1024" name="PARACHUTE_ERROR_FLAGS_GS_CONNECTION_ERROR">
+        <description>There is an error with the parachute and GS (ground station) connection</description>
```
```suggestion
        <description>There is an error with the connection between parachute and Ground Station</description>
```

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>There is an error with the parachute and GS (ground station) connection</description>
+      </entry>
+      <entry value="2048" name="PARACHUTE_ERROR_FLAGS_GPS_ERROR">
+        <description>There is an error with the parachute GPS</description>
```
```suggestion
        <description>There is an error with the parachute's GPS</description>
```

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>There is an error with the parachute GPS</description>
+      </entry>
+      <entry value="4096" name="PARACHUTE_ERROR_FLAGS_SUBSYSTEM_CONNECTION_ERROR">
+        <description>There is an error with the parachute and subsystem connection</description>
```
```suggestion
        <description>There is an error with the connection between parachute its internal subsystem</description>
```

What kind of subsystem actually? Internal to the parachute? Isn't this then just an internal error in general?

amy-chen-skydio replied:
This error is designed for different subsystems of Parachute Recovery System (PRS), it doesn't have to be internal to the parachute. For example, the subsystem could be remote controller or expansion board.

I've updated the description to `There is an error with the connection between parachute and subsystem (e.g. remote controller, expansion board, etc.)`

potaito replied:
Got it, thank you. To me a subsystem was something that's on the same device

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>There is an error with the parachute and subsystem connection</description>
+      </entry>
+      <entry value="8192" name="PARACHUTE_ERROR_FLAGS_SUBSYSTEM_FW_ERROR">
+        <description>There is an error with the parachute subsystem firmware</description>
```
What kinds of error is this referring to? Example? Wrong firmware version?

amy-chen-skydio replied:
Yes, wrong firmware version is an example of this. I've update the description to `There is an error with the parachute subsystem firmware (e.g. wrong firmware version)`

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>Reserved for future use</description>
+      </entry>
+      <entry value="65536" name="PARACHUTE_ERROR_FLAGS_LOGGING_ERROR">
+        <description>There is an error with the parachute module's logging</description>
```
```suggestion
        <description>There is an error with the parachute's internal logging system</description>
```

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>There is an error with the parachute module's logging</description>
+      </entry>
+      <entry value="131072" name="PARACHUTE_ERROR_FLAGS_MODULE_RETIRED">
+        <description>This parachute module is retired (too many deployments)</description>
```
```suggestion
        <description>This parachute module is retired. Too many deployments</description>
```

hamishwillee replied:
@potaito I'm OK with or without brackets, but if we're not using brackets we should have a proper sentence.

```suggestion
        <description>This parachute is retired (it has has been deployed too many times).</description>
```

amy-chen-skydio replied:
I've updated the description to `This parachute module is retired (i.e. too many deployments)`

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>There is an error with the parachute glow wire</description>
+      </entry>
+      <entry value="524288" name="PARACHUTE_ERROR_FLAGS_OFFBOARD_CONNECTION_ERROR">
+        <description>There is an error with the parachute and offboard computer MAVLink connection</description>
```
```suggestion
        <description>There is an error with the MAVLink connection between parachute and offboard computer</description>
```

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>There is an error with the parachute and offboard computer MAVLink connection</description>
+      </entry>
+      <entry value="1048576" name="PARACHUTE_ERROR_FLAGS_IMU_CALIBRATION_ERROR">
+        <description>Parachute module IMU calibration failed</description>
```
```suggestion
        <description>Parachute's internal IMU calibration failed</description>
```

No need to introduce "Parachute module" as another term in my opinion

hamishwillee replied:
I agree!

potaito commented on message_definitions/v1.0/common.xml:
```diff
@@ -7671,6 +7805,16 @@
+      <field type="uint8_t" name="deployment_status" enum="PARACHUTE_DEPLOYMENT_TRIGGER">Parachute deployment status</field>
+      <field type="uint8_t" name="safety_status" enum="PARACHUTE_SAFETY_FLAGS" display="bitmask">Parachute safety status</field>
+      <field type="uint8_t" name="ats_arm_altitude" units="m">Parachute Automatic Trigger System (ATS) auto-arming/disarming altitude in meters</field>
+      <field type="char[11]" name="parachute_packed_date">Parachute packed date (DD/MM/YYYY) in ASCII characters, 0 terminated. All 0: field not provided.</field>
```
Didn't we want to change this date to uint64 for less data?

If you really want to keep the string, I'd change it to YYYY-MM-DD according to [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601)

> The [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format YYYY-MM-DD (2024-05-22) is intended to harmonize these formats and ensure accuracy in all situations. Many countries have adopted it as their sole official date format, though even in these areas writers may adopt abbreviated formats that are no longer recommended.

amy-chen-skydio replied:
I've updated to use ISO 8601 format YYYY-MM-DD

potaito reviewed (commented):
Structure looks good to me. Only one comment about the date as a string still.

The rest are suggestions for cleaner phrasing.

amy-chen-skydio commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+    <enum name="PARACHUTE_DEPLOYMENT_TRIGGER">
+      <description>Parachute deployment trigger source</description>
+      <entry value="0" name="PARACHUTE_DEPLOYMENT_TRIGGER_NONE">
+        <description>None (not triggered)</description>
```
Not all the deployments are triggered automatically, such as manual trigger.

potaito replied:
So when set to "None" there is no way for the parachute to pop? It's effectively disabled then, isn't it?

amy-chen-skydio replied:
This enum is used to indicate what the trigger source was after a parachute has deployed. So "None" means that no parachute has deployed.

hamishwillee replied:
@amy-chen-skydio I didn't find this confusing, but perhaps we could improve like this?

```suggestion
        <description>None (parachute has not deployed)</description>
```

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>Drone</description>
+      </entry>
+      <entry value="4" name="PARACHUTE_DEPLOYMENT_TRIGGER_MAVLINK">
+        <description>MAVLink (offboard computer)</description>
```
@amy-chen-skydio Are you able to directly merge these suggestions? I.e. have you got a button to commit the changes above? If so, that's the best way to merge changes, because we can see what you accepted and merged and what you did not - it makes tracking changes way easier.

If you do not have that option then there is something wrong with your github repo setup. There actually is anyway, since normally I can directly edit PRs, and I can't do so for yours.

amy-chen-skydio replied:
@hamishwillee Sorry I completely missed this comment before I committed the changes. I will do this in the future.

hamishwillee replied:
Next time :-)

amy-chen-skydio committed:
Update common.xml

Address PR feedback:
 - Updated `parachute_packed_date` date string to ISO 8601 format
 - Updated descriptions for cleaner phrasing

amy-chen-skydio commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>There is an error with the parachute IMU</description>
+      </entry>
+      <entry value="4" name="PARACHUTE_ERROR_FLAGS_RF_CONNECTION_ERROR">
+        <description>There is an error with the parachute RF (used for manual control)</description>
```
Updated the wording to `There is an error with the parachute's RF that is used for manual control`

hamishwillee commented:
@potaito All your comments addressed?

potaito commented:
@hamishwillee 
Except for this one: https://github.com/mavlink/mavlink/pull/2063#discussion_r1609573608

But it's not a deal breaker, so LGTM. Thanks for the patience.

potaito reviewed (approved):
LGTM

amy-chen-skydio commented:
> @amy-chen-skydio Assuming that everyone is happy with this, we'll need similar tests/validations as for the illuminators to get this in.

@hamishwillee should I upload the tests and microservice here or should we create a PR for [mavlink-devguide](https://github.com/Skydio/mavlink-devguide) directly?

hamishwillee commented:
@amy-chen-skydio
1. Let's address @potaito final comment like this: https://github.com/mavlink/mavlink/pull/2063/files#r1626779291
2. W.r.t. the tests, could you please create the mavlink devguide entry and cross link - that way we can merge them together and you don't have to upload twice.

amy-chen-skydio committed:
Update message_definitions/v1.0/common.xml

Update `PARACHUTE_DEPLOYMENT_TRIGGER_NONE` description

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

hamishwillee commented:
Excellent. So from my perspective this is just waiting on docs  - for the tests.

amy-chen-skydio commented:
@hamishwillee here's the [PR](https://github.com/mavlink/mavlink-devguide/pull/546) for the docs and tests :pray:

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
     </enum>
+    <enum name="PARACHUTE_TRIGGER_FLAGS" bitmask="true">
+      <description>Parachute trigger sources.</description>
+      <entry value="1" name="PARACHUTE_TRIGGER_FLAGS_MANUAL">
+        <description>Manual trigger (ground based control via parachute-specific RF channel)</description>
+      </entry>
```
I think we need a more specific name for this because there are multiple manual triggers.
Note that this is intended for your remote control unit, but as described it could also be (in theory) a button on any old RC controller.

```suggestion
      <entry value="1" name="PARACHUTE_TRIGGER_FLAGS_MANUAL_RF_CHANNEL">
        <description>Manual trigger from parachute-specific RF channel</description>
      </entry>
```

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>Automatic trigger system (ATS)</description>
+      </entry>
+      <entry value="4" name="PARACHUTE_TRIGGER_FLAGS_FC">
+        <description>Flight controller trigger (e.g. MAVLink from FC, PWM, DroneCan, RC Control)</description>
```
While looking at the [docs PR](https://github.com/mavlink/mavlink-devguide/pull/546) it occurred to me that this has to play nice with the existing behaviour of [MAV_CMD_DO_PARACHUTE](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_PARACHUTE).

That command enables/disables automatic parachute triggering, or commands immediate triggering. The idea is that you perhaps want to turn off automatic triggering in some cases. While your new command allows configuration of various sources we need to consider that this would be a "master" command - i.e. if this is set, I would expect it to be the same as if `PARACHUTE_TRIGGER_FLAGS` was 0 except for "manual options", including `PARACHUTE_TRIGGER_FLAGS_MANUAL_RF_CHANNEL` and `PARACHUTE_TRIGGER_FLAGS_OFFBOARD`

There is overlap here with these flags. 

For this case

PARACHUTE_TRIGGER_FLAGS_MANUAL

What should happen if MAV_CMD_DO_PARACHUTE enables auto-triggering, but you have set PARACHUTE_TRIGGER_FLAGS_FC disabled?

What should happen if PARACHUTE_TRIGGER_FLAGS_MANUAL is

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+      <entry value="3" name="PARACHUTE_DEPLOYMENT_TRIGGER_DRONE">
+        <description>Drone</description>
+      </entry>
+      <entry value="4" name="PARACHUTE_DEPLOYMENT_TRIGGER_MAVLINK">
```
Name should be consistent with the trigger option right?

```suggestion
      <entry value="4" name="PARACHUTE_DEPLOYMENT_TRIGGER_OFFBOARD">
```

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
         <description>Illuminator thermistor failure.</description>
       </entry>
     </enum>
+    <enum name="PARACHUTE_TRIGGER_FLAGS" bitmask="true">
```
While looking at the [docs PR](https://github.com/mavlink/mavlink-devguide/pull/546) it occurred to me that this has to play nice with the existing behaviour of [MAV_CMD_DO_PARACHUTE](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_PARACHUTE).

That command enables/disables automatic parachute triggering, or commands immediate triggering. The idea is that you perhaps want to turn off automatic triggering in some cases. While your new command allows configuration of various sources we need to consider that this would be a "master" command - i.e. if this is set, I would expect it to be the same as if `PARACHUTE_TRIGGER_FLAGS` was 0 except for "manual options", including `PARACHUTE_TRIGGER_FLAGS_MANUAL_RF_CHANNEL` and `PARACHUTE_TRIGGER_FLAGS_OFFBOARD`

We need to be more precise here about what should happen if there is overlap. Thoughts?

hamishwillee replied:
More important, I think the manual trigger from [MAV_CMD_DO_PARACHUTE](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_PARACHUTE) should always be obeyed.

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -7671,6 +7805,16 @@
+      <field type="uint32_t" name="time_boot_ms" units="ms">Uptime in ms</field>
+      <field type="uint32_t" name="error_status" enum="PARACHUTE_ERROR_FLAGS" display="bitmask">Errors</field>
+      <field type="uint8_t" name="arm_status" enum="PARACHUTE_TRIGGER_FLAGS" display="bitmask">Parachute arming status</field>
+      <field type="uint8_t" name="deployment_status" enum="PARACHUTE_DEPLOYMENT_TRIGGER">Parachute deployment status</field>
```
```suggestion
      <field type="uint8_t" name="deployment_status" enum="PARACHUTE_DEPLOYMENT_TRIGGER">Parachute deployment reason/status.</field>
```

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>Auto-disarming of parachute automatic trigger system (ATS). This allows a parachute to disable ATS after detecting that it is below a desired altitude.</description>
+      </entry>
+    </enum>
+    <enum name="PARACHUTE_DEPLOYMENT_TRIGGER">
+      <description>Parachute deployment trigger source</description>
```
This is named as "trigger", but what we're reporting here is a _reason_ for deployment. So can we name it to clearly differentiate it from the trigger flags - which are "setters"? Here I have suggested "PARACHUTE_DEPLOY_REASON", because that is what it is. Need to to change everywhere.

```suggestion
    <enum name="PARACHUTE_DEPLOY_REASON">
      <description>Reason for parachute deployment (reported in PARACHUTE_STATUS)</description>
```

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+      <entry value="2" name="PARACHUTE_DEPLOYMENT_TRIGGER_ATS">
+        <description>Automatic trigger system (ATS)</description>
+      </entry>
+      <entry value="3" name="PARACHUTE_DEPLOYMENT_TRIGGER_DRONE">
```
See comment above. I have concerns about this.

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
       </entry>
     </enum>
+    <enum name="PARACHUTE_TRIGGER_FLAGS" bitmask="true">
+      <description>Parachute trigger sources.</description>
```
Mentioning the thing that uses it makes cross linking in docs a bit better.

```suggestion
      <description>Parachute trigger sources (used to enable specific trigger sources in MAV_CMD_SET_PARACHUTE_ARM).</description>
```

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+      </entry>
+    </enum>
+    <enum name="PARACHUTE_SAFETY_FLAGS" bitmask="true">
+      <description>Parachute module safety-related flags.</description>
```
```suggestion
      <description>Parachute module safety-related flags (reported in PARACHUTE_STATUS).</description>
```

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+      </entry>
+    </enum>
+    <enum name="PARACHUTE_ERROR_FLAGS" bitmask="true">
+      <description>Parachute module error flags (bitmap, 0 means no error)</description>
```
```suggestion
      <description>Parachute module error flags (bitmap, 0 means no error).  Reported in PARACHUTE_STATUS.</description>
```

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -5102,6 +5107,135 @@
+        <description>Flight Termination System (FTS) pre-checking protocol</description>
+      </entry>
+    </enum>
+    <enum name="PARACHUTE_SAFETY_FLAGS" bitmask="true">
```
I'm OK with calling this safety flags, but should we perhaps consider `PARACHUTE_STATUS_FLAGS`?

hamishwillee commented on message_definitions/v1.0/common.xml:
```diff
@@ -1873,6 +1873,11 @@
         <param index="6">Empty</param>
         <param index="7">Empty</param>
       </entry>
+      <entry value="225" name="MAV_CMD_SET_PARACHUTE_ARM" hasLocation="false" isDestination="false">
```
I wonder if this is more clear? Thoughts

```suggestion
      <entry value="225" name="MAV_CMD_SET_PARACHUTE_TRIGGERS" hasLocation="false" isDestination="false">
```
